src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+"""
+usage: workspace_downloader.py [-h] --workspace_id WORKSPACE_ID [--root_dir ROOT_DIR] [--kb_base_url KB_BASE_URL] [--workers {1,2,3,4}]
+                               [--token_filename TOKEN_FILENAME] [--delete_job_dir]
+
+PROTOTYPE - Download genome files from the workspace service (WSS).
+
+options:
+  -h, --help            show this help message and exit
+
+required named arguments:
+  --workspace_id WORKSPACE_ID
+                        Workspace addressed by the permanent ID
+
+optional arguments:
+  --root_dir ROOT_DIR   Root directory.
+  --kb_base_url KB_BASE_URL
+                        KBase base URL, defaulting to prod
+  --workers             Number of workers for multiprocessing
+  --token_filename TOKEN_FILENAME
+                        Filename in home directory that stores token
+  --delete_job_dir      Delete job directory
+
+            
+e.g.
+PYTHONPATH=. python src/loaders/workspace_downloader/workspace_downloader.py --workspace_id 39795 --kb_base_url https://ci.kbase.us/services/
+
+NOTE:
+NERSC file structure for WS:
+/global/cfs/cdirs/kbase/collections/sourcedata/ -> WS -> workspace ID -> UPA -> .fa && .meta files 
+
+e.g.
+/global/cfs/cdirs/kbase/collections/sourcedata/WS -> 39795 -> 39795_10_1 -> 39795_10_1.fa 
+                                                                         -> 39795_10_1.meta
+                                                              39795_22_1 -> 39795_22_1.fa 
+                                                                         -> 39795_22_1.meta
+                                                     
+"""
+import argparse
+import itertools
+import json
+import os
+import shutil
+import stat
+import time
+from multiprocessing import Pool, Queue, cpu_count
+
+import docker
+
+from src.clients.AssemblyUtilClient import AssemblyUtil
+from src.clients.workspaceClient import Workspace
+from src.loaders.common import loader_common_names, loader_helper
+
+# setup KB_AUTH_TOKEN as env or provide a token_filename in --token_filename
+# export KB_AUTH_TOKEN="your-kb-auth-token"
+
+# supported source of data
+SOURCE = "WS"
+# filtering applied to list objects
+FILTER_OBJECTS_NAME_BY = "KBaseGenomeAnnotations.Assembly"
+# callback_port
+CALLBACK_PORT = 9999
+
+
+class Conf:
+    def __init__(self, job_dir, output_dir, workers, kb_base_url, token_filename):
+        self.setup_callback_server_envs(job_dir, kb_base_url, token_filename)
+        time.sleep(1)
+        self.start_callback_server(docker.from_env(), "cb")
+        time.sleep(2)
+
+        # os.environ['SDK_CALLBACK_URL'] = self.cb.callback_url
+        token = os.environ["KB_AUTH_TOKEN"]
```

Comment:
This is already set in `setup_callback_server_envs`, is it needed again here?

Reply:
need token when initiate Workspace and AssemblyUtil. Already update!!

Reply:
The way the token gets set and accesses is a little roundabout. I would do something like this:

in `loader_helpers.get_token`, add the logic for getting the token from the environment and then from the file if it's not in the environment (actually, I'd swap that logic. If a user provides a token file that should take precedence).

Right at the start of `Conf.__init__`, call `get_token()`
Pass the token to `_start_callback_server` and into the Workspace and AssemblyUtil clients

Come to think of it, if you get the port at the same place you get the token, then you can pass the port into the callback server method as well and return the env and vol variables from `setup_callback_server_envs`. Then there's no need to make them instance variables

Reply:
Updated!

Reply:
:+1: